### ISSUE TITLE
ppc64le: build failure in conftab.c

### ISSUE THREAD

**Issue by gyakovlev** [ksh93/ksh#379]
continuation of https://github.com/ksh93/ksh/issues/341

```shell-session
...snip...
conftab.c:44:84: error: '_SC_ABI_AIO_XFER_MAX' undeclared here (not in a function)                                                                                                                                                      
   44 | { "ABI_AIO_XFER_MAX", { 0UL, 0 }, { 0UL, 0 }, CONF_LIMIT, CONF_C, 1, CONF_sysconf, _SC_ABI_AIO_XFER_MAX },                                                                                                                      
      |                                                                                    ^~~~~~~~~~~~~~~~~~~~                                                                                                                         
conftab.c:45:87: error: '_SC_ABI_ASYNCHRONOUS_IO' undeclared here (not in a function); did you mean '_SC_ASYNCHRONOUS_IO'?                                                                                                              
   45 | { "ABI_ASYNCHRONOUS_IO", { 0UL, 0 }, { 0UL, 0 }, CONF_LIMIT, CONF_C, 1, CONF_sysconf, _SC_ABI_ASYNCHRONOUS_IO },                                                                                                                
      |                                                                                       ^~~~~~~~~~~~~~~~~~~~~~~                                                                                                                   
      |                                                                                       _SC_ASYNCHRONOUS_IO   
conftab.c:46:80: error: '_SC_ABI_ASYNC_IO' undeclared here (not in a function); did you mean '_PC_ASYNC_IO'?
...and many other errors...
...snip...
```

full build log (built under gentoo's portage package manager)
build is running simple `sh bin/package make -S SHELL=/bin/sh`
[ksh_build_ppc64le.log.gz](https://github.com/ksh93/ksh/files/7714723/ksh_build_ppc64le.log.gz)

arch/linux.powerpc64le-64/src/lib/libast/conftab.c:
[conftab.c.gz](https://github.com/ksh93/ksh/files/7714727/conftab.c.gz)

**Comment by gyakovlev** [comment]
[ksh_build_ppc64le_clean.log.gz](https://github.com/ksh93/ksh/files/7714878/ksh_build_ppc64le_clean.log.gz)
cleaner log feed to`ansiflter` to make it more readable.

**Comment by JohnoKing** [comment]
Interestingly, I can replicate a similar build failure on x86_64 by adding `-fdiagnostics-color=always` to both `CCFLAGS` and `LDFLAGS`, e.g.:
```sh
bin/package make CC=gcc CCFLAGS='-O2 -fdiagnostics-color=always' LDFLAGS='-O2 -fdiagnostics-color=always'
```
The above command fails with many errors in `conftab.c`. Additionally, using that flag against commit f90a6745ddc7a2fcf94c5b46bdf95bef08e504a6 results in the same build error as in #341. I notice that `-fdiagnostics-color=always` is in the ppc64le build log, so that could be causing the build failure. @gyakovlev could you try removing that flag from both `CCFLAGS` and `LDFLAGS` then rebuilding on ppc64le?

**Comment by gyakovlev** [comment]
that's bizarre but yes =) apparently `-fdiagnostics-color=always` was triggering my build failures. I'm really sorry for the noise, it looked like legitimate build failure.

there are 2 test failures for 1.0.0_beta1, otherwise looks ok.

```
test pty(shcomp) begins at 2021-12-14+20:55:53                                                                                                                                                                                          
stty: 'standard input': Inappropriate ioctl for device                                                                                                                                                                                  
        shcomp-pty.ksh[164]: POSIX sh 028(C): line 179: expected "(Stopped|Suspended) \(SIGTTIN\)", got EOF                                                                                                                             
        shcomp-pty.ksh[187]: POSIX sh 029(C): line 202: expected "(Stopped|Suspended) \(SIGTTOU\)", got EOF                                                                                                                             
test pty(shcomp) failed at 2021-12-14+20:56:14 with exit code 2 [ 40 tests 2 errors ]
```
 
it even builds now without patch from https://github.com/ksh93/ksh/issues/341#issuecomment-989447964
or the [one that's merged to master](https://github.com/ksh93/ksh/pull/370)

I also tested it on ppc64 big-endian musl host, builds and passes above tests, but fails different one
```
test leaks(shcomp) begins at 2021-12-15+05:14:20                                                                                                                                                                                        
        shcomp-leaks.ksh[209]: memory leak on PATH reset before PATH search (leaked approx 12288 bytes after 2048 iterations)                                                                                                           
test leaks(shcomp) failed at 2021-12-15+05:14:23 with exit code 1 [ 41 tests 1 error ]
```

I would never suspected that flag to cause such failures.

**Comment by McDutchie** [comment]
Please try the regression tests with the latest git code from the 1.0 or master branch (or wait until I release beta 2 Real Soon Now™ if you prefer). I think those regressions may have been fixed. The memory leak one is not a real leak, 6 bytes per iteration is too small for that – it's an artefact of OS memory management.

As for `-fdiagnostics-color=always` causing build failures, that's arguably a bug, but I'm not sure it can easily be fixed. The code is probably grepping diagnostics somewhere – the AT&T folks had a liking for that. Which would be messed up by colour codes. I'll add a note to the README.

**Comment by gyakovlev** [comment]
tried building master without any package manager involvement:
```
test pty(shcomp) begins at 2021-12-15+04:45:59
        shcomp-pty.ksh[145]: POSIX sh 026(C): line 161: expected "(Stopped|Suspended)", got EOF
        shcomp-pty.ksh[169]: POSIX sh 028(C): line 184: expected "(Stopped|Suspended) \(SIGTTIN\)", got EOF
        shcomp-pty.ksh[192]: POSIX sh 029(C): line 207: expected "(Stopped|Suspended) \(SIGTTOU\)", got EOF
test pty(shcomp) failed at 2021-12-15+04:46:23 with exit code 3 [ 43 tests 3 errors ]
```
but I think it's ok to tackle it separately/later.

**Comment by McDutchie** [comment]
Do those tests only fail when run with shcomp, or also for the regular C and C.UTF-8 locale runs?

If it only fails for shcomp, I have a hard time caring. The pty tests are for the interactive shell. In fact it makes little sense to do a shcomp run for them as shcomp compiles scripts to byte code. I'll make it skip that.

**Comment by gyakovlev** [comment]
indeed, setting LANG=C.UTF-8/C changes tests results indeed ( for 1.0.0_beta1 only 1 failure)
```
stty: 'standard input': Inappropriate ioctl for device
        shcomp-pty.ksh[140]: POSIX sh 026(C): line 156: expected "(Stopped|Suspended)", got EOF
test pty(shcomp) failed at 2021-12-15+11:57:39 with exit code 1 [ 40 tests 1 error ]
```

and 1 less failure on `master`

I run tests with `LANG=C ./bin/shtests --compile`, not sure what is *run with shcomp, or also for the regular C and C.UTF-8 locale runs*

**Comment by McDutchie** [comment]
The shcomp/pty issue was fixed in 41cb2192. Since the cause of your build failure was identified and the fix (at least for now) is to document it, which will happen soon, I'm closing this issue now.
